java.lang.ArrayIndexOutOfBoundsException in org.mozilla.focus.urlinput.UrlInputPresenter$HttpRequest.get

New on 3585

Ref: https://play.google.com/apps/publish/?account=7083182635971239206#AndroidMetricsErrorsPlace:p=org.mozilla.rocket&appid=4975490230001076778&appVersion=CANARY&clusterName=apps/org.mozilla.rocket/clusters/199cfeec&detailsSpan=7

java.lang.RuntimeException: 
 
  at android.os.AsyncTask$3.done (AsyncTask.java:330)
 
  at java.util.concurrent.FutureTask.finishCompletion (FutureTask.java:354)
 
  at java.util.concurrent.FutureTask.setException (FutureTask.java:223)
 
  at java.util.concurrent.FutureTask.run (FutureTask.java:242)
 
  at android.os.AsyncTask$SerialExecutor$1.run (AsyncTask.java:255)
 
  at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1133)
 
  at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:607)
 
  at java.lang.Thread.run (Thread.java:776)
Caused by: java.lang.ArrayIndexOutOfBoundsException: 
 
  at com.android.okhttp.okio.Buffer.writeByte (Buffer.java:973)
 
  at com.android.okhttp.internal.Platform.concatLengthPrefixed (Platform.java:130)
 
  at com.android.okhttp.internal.Platform.configureTlsExtensions (Platform.java:84)
 
  at com.android.okhttp.Connection.connectTls (Connection.java:230)
 
  at com.android.okhttp.Connection.connectSocket (Connection.java:199)
 
  at com.android.okhttp.Connection.connect (Connection.java:172)
 
  at com.android.okhttp.Connection.connectAndSetOwner (Connection.java:367)
 
  at com.android.okhttp.OkHttpClient$1.connectAndSetOwner (OkHttpClient.java:130)
 
  at com.android.okhttp.internal.http.HttpEngine.connect (HttpEngine.java:329)
 
  at com.android.okhttp.internal.http.HttpEngine.sendRequest (HttpEngine.java:246)
 
  at com.android.okhttp.internal.huc.HttpURLConnectionImpl.execute (HttpURLConnectionImpl.java:457)
 
  at com.android.okhttp.internal.huc.HttpURLConnectionImpl.getResponse (HttpURLConnectionImpl.java:405)
 
  at com.android.okhttp.internal.huc.HttpURLConnectionImpl.getInputStream (HttpURLConnectionImpl.java:243)
 
  at com.android.okhttp.internal.huc.DelegatingHttpsURLConnection.getInputStream (DelegatingHttpsURLConnection.java:210)
 
  at com.android.okhttp.internal.huc.HttpsURLConnectionImpl.getInputStream (HttpsURLConnectionImpl.java)
 
  at org.mozilla.focus.urlinput.UrlInputPresenter$HttpRequest.get (UrlInputPresenter.java:123)
 
  at org.mozilla.focus.urlinput.UrlInputPresenter$QueryTask.doInBackground (UrlInputPresenter.java:98)
 
  at org.mozilla.focus.urlinput.UrlInputPresenter$QueryTask.doInBackground (UrlInputPresenter.java:86)
 
  at android.os.AsyncTask$2.call (AsyncTask.java:316)
 
  at java.util.concurrent.FutureTask.run (FutureTask.java:237)

--- a/app/src/main/java/org/mozilla/focus/urlinput/UrlInputPresenter.java
+++ b/app/src/main/java/org/mozilla/focus/urlinput/UrlInputPresenter.java
@@ -26,6 +26,7 @@ import java.lang.ref.WeakReference;
 import java.net.HttpURLConnection;
 import java.net.MalformedURLException;
 import java.net.URL;
+import java.net.URLConnection;
 import java.util.ArrayList;
 import java.util.Collections;
 import java.util.List;
@@ -39,7 +40,7 @@ public class UrlInputPresenter implements UrlInputContract.Presenter {
 
     private AsyncTask queryTask;
 
-    public UrlInputPresenter(@NonNull SearchEngine searchEngine, String userAgent) {
+    UrlInputPresenter(@NonNull SearchEngine searchEngine, String userAgent) {
         this.searchEngine = searchEngine;
         this.userAgent = userAgent;
     }
@@ -116,30 +117,15 @@ public class UrlInputPresenter implements UrlInputContract.Presenter {
 
             String line = "";
             HttpURLConnection urlConnection = null;
-            BufferedReader r = null;
+
             try {
                 urlConnection = (HttpURLConnection) url.openConnection();
                 urlConnection.setRequestProperty("User-Agent", userAgent);
-                InputStream in = new BufferedInputStream(urlConnection.getInputStream());
-                r = new BufferedReader(new InputStreamReader(in, "utf-8"));
-                StringBuilder total = new StringBuilder();
-                while ((line = r.readLine()) != null) {
-                    total.append(line).append('\n');
-                }
 
-                line = total.toString();
-            } catch (IOException | IndexOutOfBoundsException ignored) {
-                // IndexOutOfBoundsException sometimes is thrown by the okhttp library
-                // bundled within the android framework, we can only catch the exception here,
-                // or use the latest okhttp3.
+                line = readLines(urlConnection);
+            } catch (IOException ignored) {
+
             } finally {
-                if (r != null) {
-                    try {
-                        r.close();
-                    } catch (Exception e) {
-                        ;
-                    }
-                }
                 if (urlConnection != null) {
                     urlConnection.disconnect();
                 }
@@ -155,13 +141,11 @@ public class UrlInputPresenter implements UrlInputContract.Presenter {
                 JSONArray suggestions = response.getJSONArray(1);
                 int size = suggestions.length();
                 suggests = new ArrayList<>(size);
-                try {
-                    for (int i = 0; i < Math.min(size, MAX_SUGGESTION_COUNT); i++) {
-                        suggests.add(suggestions.getString(i));
-                    }
-                } catch (JSONException e) {
+
+                for (int i = 0; i < Math.min(size, MAX_SUGGESTION_COUNT); i++) {
+                    suggests.add(suggestions.getString(i));
                 }
-            } catch (JSONException e) {
+            } catch (JSONException ignored) {
             } finally {
                 if (suggests == null) {
                     suggests = Collections.emptyList();
@@ -169,6 +153,33 @@ public class UrlInputPresenter implements UrlInputContract.Presenter {
             }
 
             return suggests;
+        }
+
+        private static String readLines(URLConnection connection) throws IOException {
+            InputStream inputStream;
+            try {
+                inputStream = connection.getInputStream();
+            } catch (IndexOutOfBoundsException ignored) {
+                // IndexOutOfBoundsException sometimes is thrown by the okHttp library
+                // bundled within the android framework, we can only catch the exception here,
+                // or use the latest okHttp3.
+                return "";
+            }
+
+            StringBuilder total = new StringBuilder();
+            try (BufferedReader bufferedReader = createReader(inputStream)) {
+                String line;
+                while ((line = bufferedReader.readLine()) != null) {
+                    total.append(line).append('\n');
+                }
+            }
+
+            return total.toString();
+        }
+
+        private static BufferedReader createReader(InputStream stream) throws IOException {
+            InputStreamReader reader = new InputStreamReader(new BufferedInputStream(stream), "utf-8");
+            return new BufferedReader(reader);
         }
     }
 }

--- a/app/src/main/java/org/mozilla/focus/urlinput/UrlInputPresenter.java
+++ b/app/src/main/java/org/mozilla/focus/urlinput/UrlInputPresenter.java
@@ -128,7 +128,10 @@ public class UrlInputPresenter implements UrlInputContract.Presenter {
                 }
 
                 line = total.toString();
-            } catch (IOException e) {
+            } catch (IOException | IndexOutOfBoundsException ignored) {
+                // IndexOutOfBoundsException sometimes is thrown by the okhttp library
+                // bundled within the android framework, we can only catch the exception here,
+                // or use the latest okhttp3.
             } finally {
                 if (r != null) {
                     try {

